config/configuration.go

```diff
@@ -0,0 +1,106 @@
+package config
+
+import (
+	"os"
+	"path/filepath"
+	"runtime"
+	"strings"
+
+	"github.com/spf13/viper"
+)
+
+var (
+	// DefaultDirName is the default name used for config and workspace directories.
+	DefaultDirName string
+)
+
+// Configuration lets us inject configuration options into commands.
+// Note that we are slowly working towards getting rid of the
+// config.Config, config.UserConfig, and config.CLIConfig types.
+// Once we do, we can rename this type to Config, and get rid of the
+// User and CLI fields.
+type Configuration struct {
+	OS              string
+	Home            string
+	Dir             string
+	DefaultBaseURL  string
+	DefaultDirName  string
+	UserViperConfig *viper.Viper
+	UserConfig      *UserConfig
+	CLI             *CLIConfig
+}
+
+// NewConfiguration provides a configuration with default values.
+func NewConfiguration() Configuration {
+	home := userHome()
+
+	return Configuration{
+		OS:             runtime.GOOS,
+		Dir:            Dir(),
+		Home:           home,
+		DefaultBaseURL: defaultBaseURL,
+		DefaultDirName: DefaultDirName,
+	}
+}
+
+// SetDefaultDirName configures the default directory name based on the name of the binary.
+func SetDefaultDirName(binaryName string) {
+	DefaultDirName = strings.Replace(filepath.Base(binaryName), ".exe", "", 1)
+}
+
+// Dir is the configured config home directory.
+// All the cli-related config files live in this directory.
+func Dir() string {
+	var dir string
+	if runtime.GOOS == "windows" {
+		dir = os.Getenv("APPDATA")
+		if dir != "" {
+			return filepath.Join(dir, DefaultDirName)
+		}
+	} else {
+		dir := os.Getenv("EXERCISM_CONFIG_HOME")
+		if dir != "" {
+			return dir
+		}
+		dir = os.Getenv("XDG_CONFIG_HOME")
+		if dir == "" {
+			dir = filepath.Join(os.Getenv("HOME"), ".config")
+		}
+		if dir != "" {
+			return filepath.Join(dir, DefaultDirName)
+		}
+	}
+	// If all else fails, use the current directory.
+	dir, _ = os.Getwd()
+	return dir
+}
+
+func userHome() string {
+	var dir string
+	if runtime.GOOS == "windows" {
+		dir = os.Getenv("USERPROFILE")
+		if dir != "" {
+			return dir
+		}
+		dir = os.Getenv("HOMEDRIVE") + os.Getenv("HOMEPATH")
+		if dir != "" {
+			return dir
+		}
+	} else {
+		dir = os.Getenv("HOME")
+		if dir != "" {
+			return dir
+		}
+	}
+	// If all else fails, use the current directory.
+	dir, _ = os.Getwd()
+	return dir
+}
+
+func DefaultWorkspaceDir(cfg Configuration) string {
+	dir := cfg.DefaultDirName
+	if cfg.OS != "linux" {
```

Comment:
On Mac it probably won’t matter as the file system is case insensitive by default. I think Window is the same. At least last time I checked.

Reply:
The idiom on linux is to use lowercase directory names.

If you look at all the directories in the home directory on a mac by default, they're all capitalized (e.g. `Downloads`). Same with Windows.

So the defaults are not because the filesystems are case sensitive, but to match the pattern that they use for their other default directories.